radicle-node/src/control.rs

```diff
@@ -70,208 +76,179 @@ enum DrainError {
-) -> Result<(), DrainError> {
-    match handle.fetch(id) {
-        Err(e) => {
-            return Err(DrainError::Runtime(e));
-        }
-        Ok(FetchLookup::Found { seeds, results }) => {
-            let seeds = Vec::from(seeds);
-
-            writeln!(
-                writer,
-                "ok: found {} seeds for {} ({:?})", // TODO: Better output
-                seeds.len(),
-                &id,
-                &seeds,
-            )?;
-
-            for result in results
-                .iter()
-                .take(results.capacity().unwrap_or(seeds.len()))
-            {
-                match result.result {
-                    Ok(updated) => {
-                        writeln!(writer, "ok: {id} fetched from {}", result.remote)?;
-                        for update in updated {
-                            writeln!(writer, "{update}")?;
-                        }
-                    }
-                    Err(err) => {
-                        writeln!(
-                            writer,
-                            "error: {id} failed to fetch from {}: {err}",
-                            result.remote
-                        )?;
-                    }
-                }
-            }
-        }
-        Ok(FetchLookup::NotFound) => {
-            writeln!(writer, "error: {id} was not found")?;
+) -> Result<(), CommandError> {
+    match handle.fetch(id, node) {
+        Ok(result) => {
+            json::to_writer(&mut writer, &result)?;
         }
-        Ok(FetchLookup::NotTracking) => {
-            writeln!(writer, "error: {id} is not tracked")?;
-        }
-        Ok(FetchLookup::Error(err)) => {
-            writeln!(writer, "error: {err}")?;
+        Err(e) => {
+            return Err(CommandError::Runtime(e));
         }
     }
     Ok(())
 }
 
+mod parse {
+    use super::*;
+
+    pub(super) fn arg<T: FromStr>(cmd: Command) -> Result<T, CommandError>
+    where
+        <T as FromStr>::Err: std::error::Error + 'static,
+    {
+        let [arg]: [String; 1] = cmd
+            .args
+            .clone()
+            .try_into()
+            .map_err(|_| CommandError::InvalidCommandArgs(cmd.args))?;
+
+        arg.parse()
+            .map_err(|e| CommandError::InvalidCommandArg(arg, Box::new(e)))
+    }
+
+    pub(super) fn args<S: FromStr, T: FromStr>(cmd: Command) -> Result<(S, T), CommandError>
+    where
+        <S as FromStr>::Err: std::error::Error + 'static,
+        <T as FromStr>::Err: std::error::Error + 'static,
+    {
+        let [arg1, arg2]: [String; 2] = cmd
+            .args
+            .clone()
+            .try_into()
+            .map_err(|_| CommandError::InvalidCommandArgs(cmd.args))?;
+
+        let arg1 = arg1
+            .parse()
+            .map_err(|e| CommandError::InvalidCommandArg(arg1, Box::new(e)))?;
+        let arg2 = arg2
+            .parse()
+            .map_err(|e| CommandError::InvalidCommandArg(arg2, Box::new(e)))?;
+
+        Ok((arg1, arg2))
+    }
+}
```

Comment:
Maybe we could generalize this by using a parser? Either as a function or a type. This could simplify the code above and generalize it.  So:

```
let cmd: Command = json::from_str(input)?;
let mut parser = parser(cmd);
...
        CommandName::Fetch => {
            let rid: Id = parser.parse()?;
            let nid: NodeId = parser.parse()?;
```

or something like:

```
        CommandName::Fetch => {
            let (rid, args) = parseArgs(cmd.args)?;
            let (nid, args) = parseArgs(args)?;
```

Reply:
We could, but I think it's not worth it until there are more cases.